modules/javafx.graphics/src/main/native-font/MacFontFinder.c

```diff
@@ -193,4 +196,72 @@ JNIEXPORT jobjectArray JNICALL Java_com_sun_javafx_font_MacFontFinder_getFontDat
     return result;
 }
 
+JNIEXPORT jobjectArray JNICALL
+Java_com_sun_javafx_font_MacFontFinder_getCascadeList
+(JNIEnv *env, jclass cls, jlong fontRef)
+{
+    CTFontRef ctFontRef = (CTFontRef)fontRef;
+
+    CFArrayRef codes = CFLocaleCopyISOLanguageCodes();
+    CFArrayRef fds = CTFontCopyDefaultCascadeListForLanguages(ctFontRef, codes);
+    CFRelease(codes);
+
+    CFIndex cnt = CFArrayGetCount(fds);
+    jclass jStringClass = (*env)->FindClass(env, "java/lang/String");
+    int SPE = 2;
+    jobjectArray names = (*env)->NewObjectArray(env, cnt*SPE, jStringClass, NULL);
+    if (names == NULL) {
+        CFRelease(fds);
+        return NULL;
+    }
+    for (CFIndex i=0; i<cnt; i++) {
+        CTFontDescriptorRef ref = CFArrayGetValueAtIndex(fds, i);
+        CFStringRef fontname = CTFontDescriptorCopyAttribute(ref, kCTFontNameAttribute);
+        CFStringRef displayName = CTFontDescriptorCopyAttribute(ref, kCTFontDisplayNameAttribute);
+        CFURLRef url = CTFontDescriptorCopyAttribute(ref, kCTFontURLAttribute);
+        CFStringRef file = url ? CFURLCopyFileSystemPath(url, kCFURLPOSIXPathStyle) : NULL;
+
+        jstring jFontDisplayName = createJavaString(env, displayName);
+        CFRelease(displayName);
+        jstring jFile = createJavaString(env, file);
+        if (file != NULL) CFRelease(file);
+
+        (*env)->SetObjectArrayElement(env, names, (i*SPE)+0, jFontDisplayName);
+        (*env)->SetObjectArrayElement(env, names, (i*SPE)+1, jFile);
+
+        (*env)->DeleteLocalRef(env, jFontDisplayName);
+        if (jFile != NULL) (*env)->DeleteLocalRef(env, jFile);
+    }
+    CFRelease(fds);
+    return names;
+}
+
+JNIEXPORT jlongArray JNICALL
+Java_com_sun_javafx_font_MacFontFinder_getCascadeListRefs
+(JNIEnv *env, jclass cls, jlong fontRef)
+{
+    CTFontRef ctFontRef = (CTFontRef)fontRef;
+
+    CFArrayRef codes = CFLocaleCopyISOLanguageCodes();
+    CFArrayRef fds = CTFontCopyDefaultCascadeListForLanguages(ctFontRef, codes);
+    CFRelease(codes);
+
+    CFIndex cnt = CFArrayGetCount(fds);
+    jlongArray refs = (*env)->NewLongArray(env, cnt);
+    if (refs == NULL) {
+        CFRelease(fds);
+        return NULL;
+    }
+    jlong *refArr = calloc(cnt, sizeof(jlong));
```

Comment:
I recommend checking the return value for null.

Reply:
done